src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java

```diff
@@ -299,14 +374,113 @@ protected int doWork() {
+
+        final IntervalList regionOfInterest = getIntervals(sequenceDictionary);
+
+        log.info("Getting SamLocusIterator");
+
+        final SamLocusIterator samLocusIterator = new SamLocusIterator(sam, regionOfInterest);
+
+        // Now go through and compare information at each of these sites
+        samLocusIterator.setEmitUncoveredLoci(false);
+        samLocusIterator.setMappingQualityScoreCutoff(MIN_MAPPING_Q);
+        samLocusIterator.setQualityScoreCutoff(MIN_BASE_Q);
+
+        log.info("Using " + aggregatorList.size() + " aggregators.");
+
+        aggregatorList.forEach(la ->
+                IOUtil.assertFileIsWritable(new File(OUTPUT + la.getSuffix())));
+
+        // iterate over loci
+        log.info("Starting iteration over loci");
+
+        final SamLocusAndReferenceIterator iterator = new SamLocusAndReferenceIterator(referenceSequenceFileWalker, samLocusIterator);
+
+        // This hasNext() call has side-effects. It loads up the index and makes sure that
+        // the iterator is really ready for
+        // action. Calling this allows for the logging to be more accurate.
+        iterator.hasNext();
+        return iterator;
+    }
+
+    /**
+     * Initializes the counts, random object, and iterators for this class.
+     *
+     * Specifically, the following fields are initialized:
+     *
+     *   random
+     *   aggregatorList
+     *   progressLogger
+     *   nTotalLoci
+     *   nSkippedLoci
+     *   nProcessedLoci
+     */
+    private void initializeAggregationState() {
+        random = new Random(42);
+
+        aggregatorList = getAggregatorList();
+
+        progressLogger = new ProgressLogger(log, PROGRESS_STEP_INTERVAL);
+        nTotalLoci = 0;
+        nSkippedLoci = 0;
+        nProcessedLoci = 0;
+    }
+
+
+    @Override
+    protected int doWork() {
+
+        // Initialize our iteration:
+        initializeAggregationState();
+
+        // Make sure we can read our files:
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFilesAreReadable(INTERVALS);
+
+        // Process our data based on how we will be iterating:
+        final int returnValue;
+        if ( INTERVAL_ITERATOR ) {
+            returnValue = processDataIntervalIteration();
+        }
+        else {
+            returnValue = processDataNormalIteration();
+        }
+
+        // Check if we had an error and if so, immediately return
+        // (to preserve old functionality):
+        if (returnValue != 0) {
+            return returnValue;
+        }
+
         log.info("Iteration complete, generating metric files");
 
         aggregatorList.forEach(this::writeMetricsFileForAggregator);
 
         log.info(String.format("Examined %d loci, Processed %d loci, Skipped %d loci.\n" +
                 "Computation took %d seconds.", nTotalLoci, nProcessedLoci, nSkippedLoci, progressLogger.getElapsedSeconds()));
 
-        return 0;
+        return returnValue;
```

Comment:
at this point `returnValue==0`, so why change?

Reply:
In general it's clearer to avoid hard-coded values, even simple ones.

Reply:
in that case, can you swap the if and only `return returnValue once in the end?